[PBD Interim fix] Update getExitCheckoutUrl to use the upgradeIntent for plugins and themes

andrewserong opened the pull request:
This change allows the checkout to use a `redirect_to` parameter when a user empties their cart. The `getExitCheckoutUrl` function will iterate over the `redirect_to` trying to find the deepest buried redirect, and will use that as the redirect location if a user has a valid upgrade intent. It will not follow the redirect if it is to an external URL.

When the changes to the bannerUrls for upgrading to the business plan is merged in https://github.com/Automattic/wp-calypso/pull/38203, this PR will 'switch on' the behaviour where a user who goes from the `/plugins` page to upgrade their site, and then cancels on the checkout, will be redirected back to the plugins page. The same with themes, or any individual plugin. That change is _not_ part of this PR.

#### Changes proposed in this Pull Request

* Add `upgradeIntent` and `redirectTo` to `getExitCheckoutUrl` so that a request to the checkout where a user has nothing in the cart (e.g. after they click the trash icon on their cart items) redirects to the provided `redirectTo`
* Add constants for each of the valid upgrade intents and a function for validating a provided intent against them
* Add function to parse a `redirect_to` string that includes nested redirects that may or may not be URI encoded
* Add test coverage for the new functions

When the upgrade banners are eventually updated with intent and redirect URLs the intended behaviour will look like:

![plugin-checkout-redirect-small](https://user-images.githubusercontent.com/14988353/70302621-a782b480-1851-11ea-9813-6f917f3c8c04.gif)

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

##### Test for regressions

In the below examples replace `site-name.blog` with the domain for your test site.

* Ensure that if you sign up via `/start` and reach the checkout on a paid plan, that clicking the trashcan icon in checkout redirects you to `/plans`
* From an existing site, go to add a domain via `/domains/add/site-name.blog` (you can be on a free tier and add a domain with a plan by clicking Upgrade next to a domain). Click on the trashcan icon in checkout, and you should be redirected to `/domains/add/site-name.blog`
* Go to purchase a premium theme `/themes/premium/site-name.blog` (click Show All Theme, then on the ellipsis next to a theme, select Purchase). Click on the trashcan icon in checkout, and you should be redirected to `/themes/site-name.blog`
* Create a Jetpack site via https://jurassic.ninja/ — after the site is created, set up Jetpack and connect it to WordPress.com. When you get to `wordpress.com` URLs, replace the URL with `calypso.localhost:3000/` and from the plans step select a paid plan. From the checkout screen, click the trashcan icon and you should be redirected to `/plans/my-site-name.jurassic.ninja`. From this screen go to upgrade the plan again with any of the upgrade buttons. From the checkout, click the trashcan icon, and again you should be redirected back to `/plans/my-site-name.jurassic.ninja`

##### Test this change

In the below examples replace `site-name.blog` with the domain for your test site.

* From an existing site go to `/plugins/site-name.blog` and click the upgrade nudge to upgrade to the Business plan. From the checkout, update the URL in your browser to be the same page, but include the query string `?intent=install_plugin&redirect_to=/plugins/site-name.blog` (remember to replace `site-name.blog` with the domain of your site) and reload the page.
* Click the trashcan icon and you should be redirected to `/plugins/site-name.blog`
* Repeat the above steps but use the query string `?intent=install_plugin&redirect_to=%2Fplugins%2Fsite-name.blog%3Fredirect_to%3D%2Fme` and reload the page.
* Click the trashcan icon and you should be redirected to `/me`

Fixes #38144

adamziel commented:
Just flagging here that this larger work here https://github.com/Automattic/wp-calypso/pull/38139 ditches the `upgradeIntent` state variable and moves towards using query arguments. That's not a blocker here though - we'll need one more pass to update all the state-based code to be querystring-based after everything is merged

adamziel commented:
I'm thinking about something like...

```js
import {parse} from 'url';
const isValidValue = (url : any) => typeof url === "string" && url;

parseRedirectToChain = (url, includeOriginalUrl = true) => {
    const redirectChain: string[] = [];
    if (includeOriginalUrl && isValidValue(url)) {
        redirectChain.push(url);
    }
    let currentUrl = url;
    do {
        const {query: {redirect_to}} = parse(currentUrl, true);
        if (!isValidValue(redirect_to)) {
            break;
        }
        redirectChain.push(redirect_to);
        currentUrl = redirect_to;
    } while (true);
    return redirectChain;
};

```

andrewserong commented:
Thanks for the review @adamziel! The update to use the query string slipped my mind when I was working on this. I've updated it now to use the `intent` query param, and I've added that param to the URLs in the `<Banner>` components used in the theme and plugin Business plan upgrade nudges.

In this particular case, unless I'm missing something, I don't think we need to worry about handling the `redirect_to` param. The empty checkout redirects using the Page framework so there aren't any redirect URLs that need to be generated. Or do you mean that when we bump people out of the checkout page to `/plugins` or `/themes` when they clear the cart, we want to keep a `redirect_to` param in the query string?

If we do need to handle nested `redirect_to`s, would it make sense to `encodeURIComponent` the `redirect_to` string that we want to pass around, so that in any one context there's only effectively a single `redirect_to` param? That way we mightn't need to iterate over the whole thing, but use `decodeURIComponent` when we want to extract the full URI?

E.g. a url might be `calypso.localhost:3000/checkout/my-fake-site.blog/business?intent=plugins&redirect_to=/target-url?redirect_to=%2Fanother-url%3Fredirect_to%3Dhttps%3A%2F%2Fexample.com` so effectively in this context there's only one `redirect_to`.

adamziel commented:
There's a case of user clicking on the upgrade banner on specific plugin's page, e.g. https://wordpress.com/plugins/wordpress-seo - in this case we want to go back to the same page and the `intent` query param doesn't contain enough information.

adamziel commented:
> E.g. a url might be calypso.localhost:3000/checkout/my-fake-site.blog/business?intent=plugins&redirect_to=/target-url?redirect_to=%2Fanother-url%3Fredirect_to%3Dhttps%3A%2F%2Fexample.com so effectively in this context there's only one redirect_to.

Ah sorry for not being very clear here. Your example is exactly how it works, the thing is that the redirection chain for user starting from e.g. Yoast SEO plugin page goes something like:

1. Thank you page
2. Yoast SEO plugin page

So effectively the checkout page URL becomes something like

`http://wordpress.comcheckout/mysite.wordpress.com/business?redirect_to=%2Fcheckout%2Fthank-you%2Fplans%3Fintent%3Dinstall_plugin%26redirect_to%3D%252Fplugins%252Fwordpress-seo%252Fmysite.wordpress.com%26site_launched_before_upgrade%3Dtrue`

When user empties their card, we want them to go back to Yoast SEO plugin page. However, if we just use the `redirect_to` query arg as it is, it will redirect to thank you page. We only need to use the `redirect_to` nested in `redirect_to`, that's why I mentioned nested resolving.

andrewserong commented:
Perfect, thanks @adamziel, that gives me a good use case for exploring this further!

andrewserong commented:
Just an update, thanks for the suggested function @adamziel  — I have this working locally, but it's a bit messy so I won't push the change to this PR just yet, but will finish it off tomorrow and add tests as I'm near the end of my day.

One thing I'd like to add is constants for the upgrade intents to hopefully guard against easy to make typos (https://github.com/Automattic/wp-calypso/pull/38309). Just trying to find a good place for both this and the `parseRedirectToChain`. Since the upgrade intent is kind of related to checkout, it would make sense for us to pop the upgrade intents constants in [client/lib/checkout/constants.js](https://github.com/automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/lib/checkout/constants.js#L39) and the other functions in `index.js` in that directory, or do you think upgrade intent is special enough to warrant its own library?

adamziel commented:
[client/lib/checkout/constants.js](https://github.com/automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/lib/checkout/constants.js#L39) sounds like a great place for that, thank you @andrewserong !

andrewserong commented:
@adamziel I've updated this PR with a slightly reworked version of your function, but the logic is very similar, and it's now ready for another look. I've added some test coverage, and also got it to URI decode the `redirectTo`s. I haven't changed the `lib/checkout/index.js` to Typescript, so it's plain JS for the moment, but happy to look into TS if you'd like.

I've also removed the updates to the bannerUrls since that'll be handled in https://github.com/Automattic/wp-calypso/pull/38203, so testing this PR now requires updating the URL manually from the checkout page.

adamziel commented on client/lib/checkout/index.js:
```diff
@@ -10,8 +14,61 @@ import {
+	if (
+		Array.isArray( redirectChain ) &&
+		redirectChain.length &&
+		! isExternal( redirectChain[ redirectChain.length - 1 ] )
```
Hm I've seen this responsibility in consumers of `redirect_to`, but this is actually pretty smart since it's a pretty standard check. If we need raw value, we can always introduce `getDangerousRawDeepRedirectTo` at some point in the future.

adamziel commented on client/lib/checkout/index.js:
```diff
@@ -10,8 +14,61 @@ import {
+export function getValidDeepRedirectTo( redirectTo ) {
+	const redirectChain = parseRedirectToChain( redirectTo );
+	if (
+		Array.isArray( redirectChain ) &&
```
Not blocking, but that's pretty defensive - that function always returns an array :-)
